Raise exception when a webpushd connection error occurs

bnham opened the pull request:
#### a427fa53db759bf8b1ff595b346a781de07ca84d
<pre>
Raise exception when a webpushd connection error occurs
<a href="https://bugs.webkit.org/show_bug.cgi?id=281886">https://bugs.webkit.org/show_bug.cgi?id=281886</a>
<a href="https://rdar.apple.com/138358189">rdar://138358189</a>

Reviewed by Sihui Liu.

In the past there have been some bugs which cause a WebKit client to try to talk to webpushd with
mismatching IPC protocol versions (e.g. bug 281155). This causes webpushd to terminate the IPC
connection, which then causes a default-constructed response object to be returned to the caller in
the client process.

For most webpushd IPCs, the default-constructed response makes sense. For instance, if the client
tries to send the GetPushPermissionState IPC and this type of IPC protocol mismatch occurs, then a
default-constructed PushPermissionState is returned, which is PushPermissionState::Denied. However,
for SubscribeToPushService, this does not make sense, as a default-constructed PushSubscription
object is returned which has an empty endpoint URL and keys dictionary.

Fix this by making GetPushPermissionState raise an AbortError via an AsyncReplyError if there is an
IPC protocol mismatch. I put this AsyncReplyError in the WebPushDaemonConnection header since it&apos;s a
common header imported by both _WKWebPushDaemonConnection and NetworkNotificationManager.

* Source/WebKit/NetworkProcess/Notifications/WebPushDaemonConnection.h:
* Source/WebKit/Shared/WebPushDaemonConstants.h:
* Source/WebKit/webpushd/PushClientConnection.h:
* Source/WebKit/webpushd/PushClientConnection.messages.in:
* Source/WebKit/webpushd/PushClientConnection.mm:
(WebPushD::PushClientConnection::setProtocolVersionForTesting):
* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::connectionEventHandler):
(WebPushD::WebPushDaemon::setProtocolVersionForTesting):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::(WebPushDTest, SubscribeWithBadIPCVersionRaisesExceptionTest)):
(TestWebKitAPI::(WebPushD, WKWebPushDaemonConnectionSubscribeWithBadIPCVersionRaisesException)):

Canonical link: <a href="https://commits.webkit.org/285610@main">https://commits.webkit.org/285610@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3e7ef164b082544f0c94a7c2f6946fb979ad7d3b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73090 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52519 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25898 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77306 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24328 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75205 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60324 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/301 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57446 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15928 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76157 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47460 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62903 "Found 1 new API test failure: TestWebKitAPI.HSTS.Preconnect (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37862 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44099 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20373 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22657 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65956 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20728 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78969 "Built successfully") | 
| | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/513 "Build is in progress. Recent messages:OS: Sequoia (15.0.1), Xcode: 16.0; Running checkout-pull-request; Failed to compile WebKit") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65884 "Passed tests") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/532 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62905 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65161 "Passed tests") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7153 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/368 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3106 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/397 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/383 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/397 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/Shared/WebPushDaemonConstants.h:
```diff
@@ -40,7 +40,7 @@ static constexpr Seconds silentPushTimeoutForProduction { 30_s };
 
 constexpr auto protocolVersionKey = "protocol version"_s;
-constexpr uint64_t protocolVersionValue = 4;
+constexpr uint64_t protocolVersionValue = 5;
```
Do we need to upgrade version for this change? The message format is not changed. Is it because the old code does not how to handle failure?

bnham replied:
I added a new IPC to `PushClientConnection.messages.in` and I think technically we ought to be incrementing this version any time we add an IPC message or change the order of IPC messages in this file.

I was actually thinking of also embedding the WebKit version in the IPC message (like we do for AuxiliaryProcess startup), but I guess maybe we can do that in a separate patch.

bnham committed:
Raise exception when a webpushd connection error occurs
https://bugs.webkit.org/show_bug.cgi?id=281886
rdar://138358189

Reviewed by Sihui Liu.

In the past there have been some bugs which cause a WebKit client to try to talk to webpushd with
mismatching IPC protocol versions (e.g. bug 281155). This causes webpushd to terminate the IPC
connection, which then causes a default-constructed response object to be returned to the caller in
the client process.

For most webpushd IPCs, the default-constructed response makes sense. For instance, if the client
tries to send the GetPushPermissionState IPC and this type of IPC protocol mismatch occurs, then a
default-constructed PushPermissionState is returned, which is PushPermissionState::Denied. However,
for SubscribeToPushService, this does not make sense, as a default-constructed PushSubscription
object is returned which has an empty endpoint URL and keys dictionary.

Fix this by making GetPushPermissionState raise an AbortError via an AsyncReplyError if there is an
IPC protocol mismatch. I put this AsyncReplyError in the WebPushDaemonConnection header since it's a
common header imported by both _WKWebPushDaemonConnection and NetworkNotificationManager.

* Source/WebKit/NetworkProcess/Notifications/WebPushDaemonConnection.h:
* Source/WebKit/Shared/WebPushDaemonConstants.h:
* Source/WebKit/webpushd/PushClientConnection.h:
* Source/WebKit/webpushd/PushClientConnection.messages.in:
* Source/WebKit/webpushd/PushClientConnection.mm:
(WebPushD::PushClientConnection::setProtocolVersionForTesting):
* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::connectionEventHandler):
(WebPushD::WebPushDaemon::setProtocolVersionForTesting):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::(WebPushDTest, SubscribeWithBadIPCVersionRaisesExceptionTest)):
(TestWebKitAPI::(WebPushD, WKWebPushDaemonConnectionSubscribeWithBadIPCVersionRaisesException)):

Canonical link: https://commits.webkit.org/285610@main
